execution/scan/functions.go

```diff
@@ -0,0 +1,678 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
+package scan
+
+import (
+	"math"
+
+	"github.com/prometheus/prometheus/model/histogram"
+)
+
+var invalidSample = sample{T: -1, F: 0}
+
+type sample struct {
+	T int64
+	F float64
+	H *histogram.FloatHistogram
+}
+
+type functionArgs struct {
+	Samples          []sample
+	StepTime         int64
+	SelectRange      int64
+	ScalarPoints     []float64
+	Offset           int64
```

Comment:
Is there a way to pass `Offset` and `SelectRange` via a clojure when creating the function? These values should be constant and do not change between steps, so we might get a small speed up if there are less arguments to put on the stack before the function call.

The arguments are also optional for most functions so there might be an unnecessary overhead for something like `sum_over_time`.